### PR TITLE
supports pessimistic transaction

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: proto/deadlock.proto [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	repeated WaitForEntry entries = 1 [(gogoproto.nullable) = false];
+}
+
+message WaitForEntry {
+	// The transaction id that is waiting.
+	uint64 txn = 1;
+	// The transaction id that is being waited for.
+	uint64 wait_for_txn = 2;
+	// The hash value of the key is being waited for.
+	uint64 key_hash = 3;
```
Comment (zhangjinpeng1987) [inline proto/deadlock.proto]: If there is hash conlfict, does it matter?

File: proto/deadlock.proto [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+enum DeadlockRequestType {
+	Detect = 0;
+	// CleanUpWaitFor cleans a single entry the transaction is waiting.
+	CleanUpWaitFor = 1;
+	// CleanUp cleans all entries the transaction is waiting.
+	CleanUp = 2;
```
Comment (zhangjinpeng1987) [inline proto/deadlock.proto]: Maybe CleanUpAll?

File: proto/deadlock.proto (hunk 1 above)
Comment (coocood) [reply proto/deadlock.proto]: Then it will timeout, it's very rare, so I think it's acceptable.

File: proto/kvrpcpb.proto [hunk 3]
```diff
@@ -151,13 +152,32 @@ message PrewriteRequest {
     uint64 start_version = 4;
     uint64 lock_ttl = 5;
     bool skip_constraint_check = 6;
+    // For pessimistic transaction, some mutations don't need to be locked, for example, non-unique index key.
+    repeated bool is_pessimistic_lock = 7;
```
Comment (zhangjinpeng1987) [inline proto/kvrpcpb.proto]: Embeded in Mutation is better?

Comment (coocood) [reply proto/kvrpcpb.proto]: That would increase the unmarshal overhead for the non-pessimistic transactions.

File: proto/deadlock.proto (hunk 2 above)
Comment (coocood) [reply proto/deadlock.proto]: `CleanUpAll` can be misunderstood as clean up all transactions in the detector.

File: proto/deadlock.proto (hunk 1 above)
Comment (coocood) [reply proto/deadlock.proto]: No, in this case, it doesn't matter, because there is definitely a deadlock.

Comment (coocood) [reply proto/deadlock.proto]: The key_hash can be used to make deadlock only retry a single statement.

**Comment by youjiali1995** [comment]
LGTM

File: proto/kvrpcpb.proto (hunk 3 above)
Comment (zhangjinpeng1987) [reply proto/kvrpcpb.proto]: Why?

Comment (coocood) [reply proto/kvrpcpb.proto]: the Mutation struct is bigger.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**